logic/Rule.java

```diff
@@ -203,25 +194,24 @@ private Conjunction thenPattern(ThingVariable<?> thenVariable) {
                 new Conjunction(VariableRegistry.createFromThings(list(thenVariable)).variables(), set()));
     }
 
-    public abstract static class Conclusion<CONSTRAINT extends Constraint> {
-
-        private final CONSTRAINT constraint;
-
-        private Conclusion(CONSTRAINT constraint, Set<Variable> whenContext) {
-            this.constraint = constraint;
-            copyAdditionalConstraints(whenContext, new HashSet<>(this.constraint.variables()));
-        }
-
-        public CONSTRAINT constraint() {
-            return constraint;
-        }
-
-        public static Conclusion<?> create(ThingConstraint constraint, Set<Variable> whenContext) {
-            if (constraint.isRelation()) return Relation.create(constraint.asRelation(), whenContext);
-            else if (constraint.isHas()) return Has.create(constraint.asHas(), whenContext);
-            else if (constraint.isIsa()) return Isa.create(constraint.asIsa(), whenContext);
-            else if (constraint.isValue()) return Value.create(constraint.asValue(), whenContext);
-            else throw GraknException.of(ILLEGAL_STATE);
+    public static abstract class Conclusion {
+
+        public static Conclusion create(Conjunction then) {
+            return Iterators.iterate(then.variables()).filter(Variable::isThing).map(Variable::asThing)
+                    .flatMap(variable -> Iterators.iterate(variable.constraints()).map(constraint -> {
+                        if (constraint.isRelation()) {
+                            return new Relation(constraint.asRelation(), variable.isa().get());
```

Comment:
probably good to do an assert inside each if for the right constraints existing:
```
assert variable.isa().isPresent()
```